pkg/operator/ceph/operator.go

```diff
@@ -161,11 +163,26 @@ func (o *Operator) Run() error {
 	}
 }
 
-func (o *Operator) startSystemDaemons(clusterSpec *cephv1.ClusterSpec) error {
+func (o *Operator) startDrivers() error {
 	if o.delayedDaemonsStarted {
 		return nil
 	}
 
+	o.delayedDaemonsStarted = true
```

Comment:
Why not set this to true at the end of the method? Then you don't need to set it to false if it fails

Reply:
This is to let the updateDriver know that call came through startDriver.
We need this check to block the updateDriver from running if there is no CSI driver already started.
If we create configmap before we create a CephCluster, the updateDriver would start the CSI driver which isn't nice. With this, it will wait for someone to call startDriver before it starts updating.

Reply:
Wait, the `delayedDaemonsStarted` doesn't represent whether or not the CSI driver should be started. There are other settings that indicate if the rbd or cephfs drivers should be started. Those settings also might change when the configmap is updated, so we need to always call `updateDrivers()` when the configmap is updated, right? The check on line 182 shouldn't be necessary.

Reply:
Here's the scenario.
1. I install the operator
2. I create the ConfigMap
    2.1 Watch detects it and calls UpdateDriver
    2.2 Update driver will start CSI driver
3. I create my first Ceph Cluster
    3.1 I get logs saying CSI is already started and creation proceeds
4. Everything works fine hereafter

Now, the question is "Should we be starting CSI just because ConfigMap was created? Or, should we wait for the first Ceph Cluster to be created?"

IIRC, you said `delayedDaemonStarted` tracks if CSI driver already started (for some cluster). So, I used it to block anyone from directly calling UpdateDriver, without first calling StartDriver (which means only Ceph Cluster creation can do it).

So, after this the flow becomes,
1. I install the operator
2. I create the ConfigMap
    2.1 Watch detects it and calls UpdateDriver
    2.2 Update Driver will see that CSI driver is not enabled yet and return.
3. I create my first Ceph Cluster
    3.1 It will call Start Driver and set `delayedDaemonStarted`
    3.2 Then it will call Update Driver which sees that CSI Driver is enabled and updates.
4. Now if I update Config Map,
   4.1 watch will pick up update and call Update Driver
   4.2 Update Driver will see that `delayedDaemonStarted` is true meaning CSI is present
   4.3 It will begin update and proceed further

Reply:
Ah, thanks for explaining the scenario. A couple suggestions/questions:
- The comment on line 181 should say something like "Skipping CSI driver config since the first cluster hasn't been started yet"
- We need to ignore events from the configmap watcher until after the first cluster is started. Do we need a lock on this method in case the configmap is updated during the first cluster creation? Maybe that's overkill though.

Reply:
>     * We need to ignore events from the configmap watcher until after the first cluster is started. Do we need a lock on this method in case the configmap is updated during the first cluster creation? Maybe that's overkill though.
>
We don't need to ignore as the number of events is going to be ~1 .
We don't need to lock it as no matter who calls `updateDriver` second, the specs are going to be the same and update won't actually proceed.
And if config updates while cluster is getting created, csi pods can independently restart anyway.